apollo-ios-codegen/Sources/GraphQLCompiler/GraphQLSchema.swift

```diff
@@ -239,16 +239,20 @@ public final class GraphQLObjectType: GraphQLCompositeType, GraphQLInterfaceImpl
   public private(set) var fields: [String: GraphQLField]!
 
   public private(set) var interfaces: [GraphQLInterfaceType]!
+  
+  public private(set) var keyFields: [String]!
```

Comment:
I made `keyFields` required, just to make sure the data is loaded correctly from the Javascript environment. The same field is Optional in the generated code.

Reply:
```gql
extend interface Node @typePolicy(keyFields: "id")
```

Every type I want to use in my specific iOS project has a base interface with the ID field, so it's important to me that the directive works on interfaces.

If a GQL server introduces a new type for an old endpoint used by a Client app that isn't up to date, it will now just miss the directive because the Object declaration doesn't exist. It's still possible to cache it anyway using Programmatic Cache Keys.

Reply:
Thinking about this more, it seems more complicated than we realized. We are in the process of implementing a feature to reduce the number of generated object types. Specifically because of this common practice of using a `Node` interface. We've had complaints from users of massive generated schemas that are problematic for binary size.

We intend on removing the generation of all of the object types for every type that implements the referenced interfaces. Unless the concrete types or directly referenced themselves, we don't want to generate all of them anymore. 

In this situation, types that are not referenced directly will behave the same way as "unknown types" (types added to the schema after code-gen). This makes these edge cases a LOT more likely to appear and means that the approach we're using here of adding `keyFields` to only the object types will likely not be sufficient.

We are definitely going to need to add the `keyFields` to the interface types as well. In most cases, when the object type does not exist, we could infer the `@typePolicy` from the interface type for a field. But this still winds up causing a lot of complications that I'm not quite sure how to address. It sounds like in Apollo Kotlin we aren't really handling those complications right now. But I'd like us to try and come up with an approach to those issues here.

1. What if a type that does implement an interface with a `@typePolicy` is returned for a field of another interface that does not have a `@typePolicy`? We wouldn't know that it has an appropriate `@typePolicy` to use at run time.

2. What if a type is returned from two fields with different interface type that  have a conflicting `@typePolicy` directives?

Reply:
The `Node` interface seems to originate from this "best practice" page on the GraphQL website: https://graphql.org/learn/global-object-identification/

Reply:
This came up in the Kotlin repo today: https://github.com/apollographql/apollo-kotlin/issues/6318

Reply:
If inferring a typePolicy is desired, it would become its own PR. I made the issue on the Kotlin repo so the discussion wouldn't block this Pull Request.

Since the `Node` interface is part of the Relay Specification, it's a commonly encountered pattern. However, since the Client app doesn't have any information for a result it doesn't recognize, it can't be 100% sure if it implements the interface or not. I would consider swapping out the result type of an old query to be a rare breaking change anyway, so not caching this new unknown object seems like a reasonable default.

Reply:
Ah! I didn't realize you were the same `@x-sheep`, sorry 😄 Thanks for opening this 🙏

Reply:
I thought of an example case (ignoring typePolicy overrides for a moment). If this schema was fetched through introspection at compile time:

```gql
interface Node {
    id: ID!
}

type OldType implements Node {
    id: ID!
    field: String
}

type Query {
    someTypes: [OldType]
}
```

And this query is used:

```gql
extend interface Node @typePolicy(keyFields: "id")

query GetTypes {
    someTypes {
        __typeName
        field
    }
}
```

And at some point in the future the schema is suddenly changed to this:

```gql
type OldType implements Node @deprecated {
    id: ID!
    field: String
}

type NewType {
    id: ID! # Oops, no interface
    field: String
    otherField: String
}

type Query {
    someTypes: [NewType]
}
```

I would be fairly surprised if a cache started filling up with `"NewType:*"` keys, but one new schema fetch and introspection later it _stops_ inserting/hitting cache items because it knows the type has changed but the policy hasn't changed with it. 

Changing the schema in this way isn't technically a _breaking change_ according to the GraphQL spec (which defines it as a previously valid request becoming invalid). So whichever assumption this PR ends up with will lead to one of these outcomes:

1. Caching items that shouldn't have been cached
2. Not caching items that should've been cached (the PR is currently here)

Reply:
I think I agree with you that we shouldn't be computing cache keys for `NewType` in that situation (though I'm not convinced that this isn't considered a breaking change.

But I'm more concerned about this situation:

```gql
interface CachableNode @typePolicy(keyFields: "id") {
    id: ID!
}

type OldType implements CachableNode {
    id: ID!
    field: String
}

type Query {
    someTypes: [CachableNode!] // <-- Interface type instead of concrete OldType
}
```

In this case, how should we handle `NewType` when it starts being returned by the server at some point in the future?

Reply:
This is definitely a breaking change:

```diff
type Query {
-    someTypes: [OldType]
+    someTypes: [NewType]
}
```

as that (old) query doesn't validate anymore:

```graphql
{
  someTypes {
    ... on OldType { __typename }
  }
}
```

Agree with @AnthonyMDev's question. I _think_ if we forbid overriding the type policy we can support new fields without recompilation but maybe I'm missing a weird case

Reply:
> `someTypes: [CachableNode!] // <-- Interface type instead of concrete OldType`

That does help me understand the issue with interfaces, thank you. I think the current implementation of Cache key calculation doesn't support interfaces, it either comes with a concrete type (with interface information) or nothing at all, so even programmatic cache keys don't know anything about the request itself. The interface `@typePolicy` isn't a new problem, in that sense.

```swift
// apollo-ios/Sources/ApolloAPI/SchemaMetadata.swift
public static func cacheKey(for object: ObjectData) -> String? {
    /* object["__typeName"] is the only metadata available here */
}
```

Adding the query-time expected interface to this call signature would be a considerable change, which I think is better as a separate PR.

Reply:
> Adding the query-time expected interface to this call signature would be a considerable change, which I think is better as a separate PR.

I don't think this is needed? Cache keys should always contain concrete typename and this is fine:

```graphql
{
  node(id: "42") {
    id
    ... on Product {
      price
      name
    }
  }
}
```

If a product is returned, its Cache key is `Product:42` and this is fine? If a Record references the product, it'll have a `CacheReference("Product:42")`. No need for interface information?

Reply:
The problem isn't that the type name is unknown, but that it's not certain if a key _should_ be generated at all. The only way to make the above work (in current code) is to unconditionally use `id` as the key:

```swift
static func cacheKeyInfo(for type: Object, object: ObjectData) -> CacheKeyInfo? {
  guard let id = object["id"] as? String else {
      return nil
  }
  return CacheKeyInfo(id: id) // CachableNode interface or not?
}
```

Reply:
Right. You'd need to know that `new object` is an instance of `CacheableNode` in `cacheKeyInfo`. 

I think this works in Apollo Kotlin because we codegen keyfields for interfaces as well.

https://github.com/apollographql/apollo-kotlin/blob/7668a302eb5c62ca1d3e423ccfefa1142acee1f4/libraries/apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/api/CacheKeyGenerator.kt#L52:

```kotlin
object TypePolicyCacheKeyGenerator : CacheKeyGenerator {
  override fun cacheKeyForObject(obj: Map<String, Any?>, context: CacheKeyGeneratorContext): CacheKey? {
    val keyFields = context.field.type.rawType().keyFields() // this comes from codegen

    return if (keyFields.isNotEmpty()) {
      // We might not know the type at build time but we know at runtime it has keyfields from CacheableNode
      CacheKey(obj["__typename"].toString(), keyFields.map { obj[it].toString() })
    } else {
      null
    }
  }
}
```

Reply:
Should this be addressed in a separate PR? If yes, should it be done before or after merging this `@typePolicy` branch?

Reply:
**I think we do need to add the `keyFields` to interface types and add that to the handling of keyFields for a type that is not a known generated type.** 

Even if we aren't handling all the cases of "unknown types" this is important because we are currently building a feature to allow many concrete types to not be generated (if they aren't directly referenced, but only can be returned for a field from an interface).

So this means that ALOT of known types would not normalize properly as well, as they would appear at run-time to be unrecognized types.

---

I really would have loved for us to be able to solve the issues with handling unknown types here in a more elegant way, but for now, I think this is what we need to do. Which, to be fair, is actually just mirroring what Apollo Kotlin does currently, and they haven't seen a lot of issues come up on that side. The only thing that's not working properly here is handling cache normalization for an unknown (or not-generated) type that, conforms to two interfaces with conflicting `@typePolicy` directives. But it's a very rare, minor edge case, and there isn't a great way to fix for that right now. So I think that we can't let that hold up pushing this feature.

Reply:
+1. Also about this part:

> The only thing that's not working properly here is handling cache normalization for an unknown (or not-generated) type that, conforms to two interfaces with conflicting @typePolicy directives.

Longer term, we could add linting/schema checks to catch that before the schema is actually put in production.

Reply:
This PR already throws an error if any type is found that has conflicting typePolicies through its inherited interfaces. As soon as the schema is updated, any new conflict will become apparent. I'm not aware of what the changes look like when Apollo iOS generates less code, but if the entire schema is still validated in the Javascript/IR step this will still catch conflicting policies (regardless of their usage by the client).

I can update this PR to add typePolicy information to the Interfaces as well as the Objects, so it will also run the above validation on interfaces without (known) object types.

Reply:
I've now added keyFields to the generated Interface instances.